### PR TITLE
Generalize norms

### PR DESCRIPTION
**Pull request by IainNZ**

### CODE REVIEW

File: src/v0.4/parseExpr_staged.jl [hunk 1]
```diff
@@ -398,14 +398,22 @@ function parseNorm(x::Expr, aff::Symbol, lcoeffs, rcoeffs, newaff)
...
-            $normexpr = AffExpr[]
+            $normexpr = GenericAffExpr[]
             _sizehint_expr!($normexpr, $len)
         end
     end
+    param = 2
+    if normp == :norm1
+        param = 1
+    elseif normp ∈ [:norminf, :norm∞]
+        param = -1
```
Comment (joehuchette) [inline src/v0.4/parseExpr_staged.jl]: :Infinity?

Comment (joehuchette) [inline src/v0.4/parseExpr_staged.jl]: Or just ``Inf``

File: src/v0.4/parseExpr_staged.jl [hunk 2]
```diff
@@ -383,10 +383,10 @@ function parseNorm(x::Expr, aff::Symbol, lcoeffs, rcoeffs, newaff)
                 $code
             end)
         end
-        preblock = :($normexpr = AffExpr[])
+        preblock = :($normexpr = GenericAffExpr[])
```
Comment (joehuchette) [inline src/v0.4/parseExpr_staged.jl]: I'm not sure this'll work when you try to construct the ``Norm``, and I wonder if there will be performance hits.

File: src/v0.4/parseExpr_staged.jl (hunk 1 above)
Comment (IainNZ) [inline src/v0.4/parseExpr_staged.jl]: Oh didn't know you could mix it up

**Comment by IainNZ** [comment]
# This PR

### Julia 0.3.7
```
WARNING: The addition operator has been used on JuMP expressions a large number of times. For performance reasons, you should not add expressions in a loop. Instead of x += y, use append!(x,y) to modify x in place. If y is a single variable, you may also use push!(x, coef, y) in place of x += coef*y.
elapsed time: 2.615578351 seconds (616210084 bytes allocated, 11.23% gc time)
```
### Julia 0.4.0-dev+6540 (2015-08-08 20:25 UTC)
```
WARNING: The addition operator has been used on JuMP expressions a large number of times. For performance reasons, you should not add expressions in a loop. Instead of x += y, use append!(x,y) to modify x in place. If y is a single variable, you may also use push!(x, coef, y) in place of x += coef*y.
  0.142680 seconds (1.85 M allocations: 220.805 MB, 13.11% gc time)
```

# Master

### Julia 0.3.7
```
WARNING: The addition operator has been used on JuMP expressions a large number of times. For performance reasons, you should not add expressions in a loop. Instead of x += y, use append!(x,y) to modify x in place. If y is a single variable, you may also use push!(x, coef, y) in place of x += coef*y.
elapsed time: 0.26372442 seconds (224318832 bytes allocated, 36.79% gc time)
```
### Julia 0.4.0-dev+6540 (2015-08-08 20:25 UTC)
```
WARNING: The addition operator has been used on JuMP expressions a large number of times. For performance reasons, you should not add expressions in a loop. Instead of x += y, use append!(x,y) to modify x in place. If y is a single variable, you may also use push!(x, coef, y) in place of x += coef*y.
  0.159656 seconds (2.05 M allocations: 227.086 MB, 14.50% gc time)
```

File: src/v0.3/parseExpr_0.3.jl [hunk 3]
```diff
@@ -287,7 +287,7 @@ function parseNorm(x::Expr, aff::Symbol, coeffs)
         newaff, innercode = parseExprToplevel(x.args[3], :normaff)
         code = quote
             if $(esc(cond.args[1]))
-                normaff = zero(AffExpr)
+                normaff = 0.0
```
Comment (joehuchette) [inline src/v0.3/parseExpr_0.3.jl]: Why the change?

File: src/v0.3/parseExpr_0.3.jl [hunk 4]
```diff
@@ -298,10 +298,10 @@ function parseNorm(x::Expr, aff::Symbol, coeffs)
                 $code
             end)
         end
-        preblock = :($normexpr = AffExpr[])
+        preblock = :($normexpr = GenericAffExpr[])
     else # no condition
         newaff, code = parseExprToplevel(x.args[2], :normaff)
-        code = :(normaff = zero(AffExpr); $code; push!($normexpr, $newaff))
+        code = :(normaff = 0.0; $code; push!($normexpr, $newaff))
```
Comment (joehuchette) [inline src/v0.3/parseExpr_0.3.jl]: Ditto

**Comment by joehuchette** [comment]
v0.3:
```
julia> macroexpand(:( @addConstraint(m, norm2{x[j],j=1:i} <= i)))
quote  # /Users/huchette/.julia/v0.3/JuMP/src/macros.jl, line 149:
    JuMP.validmodel(m,:m) # line 150:
    begin  # /Users/huchette/.julia/v0.3/JuMP/src/macros.jl, line 296:
        #1538#q = JuMP.zero(JuMP.AffExpr) # line 297:
        begin
            begin
                begin  # /Users/huchette/.julia/v0.3/JuMP/src/v0.3/parseExpr_0.3.jl, line 330:
                    begin  # /Users/huchette/.julia/v0.3/JuMP/src/v0.3/parseExpr_0.3.jl, line 314:
                        #1533###24092 = 0 # line 315:
                        for j = 1:i
                            #1533###24092 += JuMP.length(1:i)
                        end # line 316:
                        #1534###24093 = JuMP.GenericAffExpr[] # line 317:
                        JuMP._sizehint_expr!(#1534###24093,#1533###24092)
                    end # line 331:
                    for j = 1:i # line 309:
                        begin
                            #1535#normaff = 0.0
                            begin
                                #1536###24094 = JuMP.addToExpression(#1535#normaff,(*)(1.0),x[j])
                                JuMP.push!(#1534###24093,#1536###24094)
                            end
                        end
                    end # line 332:
                    #1537###24091 = JuMP._build_norm(2,#1534###24093) # line 333:
                    #1538#q = JuMP.addToExpression(#1538#q,(*)(1.0),#1537###24091)
                end
                #1539###24095 = JuMP.addToExpression(#1538#q,-1.0 * (*)(1.0),i)
            end
            #1540###24088 = #1539###24095
        end # line 298:
        #1541###24087 = JuMP.addConstraint(m,JuMP.constructconstraint!(#1540###24088,:<=))
    end
end
```
v0.4:
```
julia> macroexpand(:( @addConstraint(m, norm2{x[j],j=1:i} <= i)))
quote  # /Users/huchette/.julia/v0.4/JuMP/src/macros.jl, line 149:
    JuMP.validmodel(m,:m) # line 150:
    begin  # /Users/huchette/.julia/v0.4/JuMP/src/macros.jl, line 296:
        #1#q = JuMP.zero(JuMP.AffExpr) # line 297:
        begin
            begin  # /Users/huchette/.julia/v0.4/JuMP/src/v0.4/parseExpr_staged.jl, line 415:
                begin  # /Users/huchette/.julia/v0.4/JuMP/src/v0.4/parseExpr_staged.jl, line 399:
                    #2###6495 = 0 # line 400:
                    for j = 1:i
                        #2###6495 += JuMP.length(1:i)
                    end # line 401:
                    #3###6496 = JuMP.GenericAffExpr[] # line 402:
                    JuMP._sizehint_expr!(#3###6496,#2###6495)
                end # line 416:
                for j = 1:i # line 394:
                    begin
                        #4#normaff = 0.0
                        begin
                            #5###6497 = JuMP.addToExpression_reorder(#4#normaff,x[j])
                            JuMP.push!(#3###6496,#5###6497)
                        end
                    end
                end # line 417:
                #6###6494 = JuMP._build_norm(2,#3###6496) # line 418:
                #7###6492 = JuMP.addToExpression_reorder(#1#q,#6###6494)
            end
            #8###6491 = JuMP.addToExpression_reorder(#7###6492,-1.0,i)
        end # line 298:
        #9###6490 = JuMP.addConstraint(m,JuMP.constructconstraint!(#8###6491,:(<=)))
    end
end
```
I'm guessing it's the
```
#1538#q = JuMP.zero(JuMP.AffExpr)
...
#1538#q = JuMP.addToExpression(#1538#q,(*)(1.0),#1537###24091)
```
bit that's hurting on v0.3.

**Comment by mlubin** [comment]
I'm not too concerned about performance of norms on 0.3, as long as we don't introduce performance regressions for linear and quadratic constraints.

File: src/JuMP.jl [hunk 5]
```diff
@@ -543,6 +543,13 @@ end
 
 typealias Norm{Typ} GenericNorm{Typ,Float64,Variable}
 
+_build_norm(Lp, terms::Vector{GenericAffExpr}) = _build_norm(Lp, [terms...])
```
Comment (joehuchette) [inline src/JuMP.jl]: Isn't splatting a long-ish vector a bad idea?

**Comment by joehuchette** [comment]
LGTM, besides the splatting thing.

**Comment by IainNZ** [comment]
Cool. Yeah I'll check the splatting thing.
Also, I'm adding more to this PR, widening some operator overloads.

**Comment by IainNZ** [comment]
OK I've widened the norm stuff, I can reuse almost everything in JuMPeR now with very little extra work.
I've also fixed the performance regression on 0.3 - it was actually the splatting, which works perfectly fine in 0.4 but sucks in 0.3 apparently.
